### PR TITLE
Slash Commands

### PR DESCRIPTION
**Pull request by binaryoverload**

### CODE REVIEW

File: src/main/kotlin/org/cascadebot/bot/cmd/DashboardCommand.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import org.cascadebot.bot.cmd.meta.CommandContext
+import org.cascadebot.bot.cmd.meta.RootCommand
+
+class DashboardCommand : RootCommand("dashboard", "Provides a link to login to the dashboard"){
+
+    override val commandData: CommandData = Commands.slash(name, description)
+
+    override fun onCommand(context: CommandContext, args: CommandArgs, data: Any) {
+//        context.reply("HI")
+        context.replyE("Test")
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/cmd/DashboardCommand.kt]: **E**

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/cmd/DashboardCommand.kt]: replyE method doesn't even exist lol

File: src/main/kotlin/org/cascadebot/bot/events/ReadyListener.kt [hunk 2]
```diff
@@ -23,5 +24,7 @@ class ReadyListener(): ListenerAdapter() {
             rabbitMQManager.channel.queueDeclare("shard-$shardId", true, false, false, mapOf())
             rabbitMQManager.channel.queueBind("shard-$shardId", "amq.topic", "shard.$shardId.*.#")
         }
+
+        Main.commandManager.registerCommandsOnce(event.jda)
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/events/ReadyListener.kt]: We should probably talk about this, considering we're barely going to have any global commands I suppose it's fine, just feels wrong.

File: src/main/kotlin/org/cascadebot/bot/events/InteractionListener.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        val args = CommandArgs(event.options.groupBy { it.name })
+        if (event.isGlobalCommand) {
+            logger.info(event.commandPath)
+            logger.info(Main.commandManager.commands.map { it.first.path.joinToString(", ", "[", "]")}.joinToString { "\n" })
+            Main.commandManager.getCommand(event.commandPath)?.onCommand(
+                CommandContext(event),
+                args,
+                ""
+            )
+        }
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/events/InteractionListener.kt:39]: we should probably throw an exception in the else of this if. Theoretically it should never happen, but if it does we should know about it.

File: src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import kotlin.system.exitProcess
+import kotlin.system.measureTimeMillis
+
+class CommandManager {
+
+    private val _commands: MutableMap<CommandPath, ExecutableCommand> = mutableMapOf()
+    val commands
+        get() = _commands.toList()
+
+    private val registeredCommands: AtomicBoolean = AtomicBoolean(false)
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt:21]: This never gets updated

**Review by weeryan17 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+        // Already registered commands on another shard
+        if (registeredCommands.getAndSet(true)) return logger.debug("Commands have already been registered on another shard")
+
+        var successful = 0
+        val elapsedTime = measureTimeMillis {
+            for ((path, command) in _commands.mapKeys { it.key.path.joinToString("/") }) {
+                logger.debug("Attempting to register command '$path'")
+                try {
+                    val cmd = jda.upsertCommand(command.commandData).complete()
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt]: Ok I know we're not going to upload many commands, but I fell like complete is wrong to do. We should probably have a quick conversation about this

File: src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    fun registerCommandsOnce(jda: JDA) {
+        logger.debug("Beginning to register commands with Discord")
+
+        // Already registered commands on another shard
+        if (registeredCommands.getAndSet(true)) return logger.debug("Commands have already been registered on another shard")
+
+        var successful = 0
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt]: Why does this variable exist? We (currently) don't use it anywhere

Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt]: We do! See lines 63 and 69

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt]: Oh I see you use it for the log, missed that my bad.

File: src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt (hunk 5 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt]: I've changed it to use queue, thoughts?

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt]: looks good, neat that you can combine rest actions like that.

**Review by weeryan17 — APPROVED** [review]

**Pull request merged**